File: BFS.java
import java.util.*;
import java.util.LinkedList;
import java.io.*;

public class BFS {

    static void bfs(ArrayList<ArrayList<Integer>> edges, int startNode) {
        ArrayList<Boolean> visited = new ArrayList<>(edges.size());
        for (int i = 0; i < edges.size(); i++) {
            visited.add(false);
        }
        Queue<Integer> queue = new LinkedList<Integer>();

        queue.add(startNode);
        visited.set(startNode, true);
        while (!queue.isEmpty()) {
            int u = queue.poll();
            System.out.print(u + " ");
            for (int v : edges.get(u)) {
                if (!visited.get(v)) {
                    queue.add(v);
                    visited.set(v, true);
                }
            }
        }
    }

    public static void main(String[] args) {
        // System.out.println("Hello World!!!");
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        ArrayList<ArrayList<Integer>> edges = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            edges.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            edges.get(u).add(v);
            edges.get(v).add(u);
        }
        int startNode = 0;
        bfs(edges, startNode);
    }
}

// 5 6
// 0 1
// 0 2
// 1 3
// 1 2
// 4 3
// 4 2

File: bfs.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <list>
#include <queue>
#include <set>
#include <string>
#include <vector>

using namespace std;

typedef long long ll;
#define ln "\n"
#define mp make_pair
#define pb push_back
#define f first
#define s second
#define INF 2e18

void bfs(vector<vector<int>> &edges, int startNode) {
	vector<bool> visited(edges.size(), false);
	queue<int> q;
	q.push(startNode);
	visited[startNode] = true;
	while (!q.empty()) {
		int u = q.front();
		q.pop();
		cout << u << " ";
		for (int v : edges[u]) {
			if (!visited[v]) {
				q.push(v);
				visited[v] = true;
			}
		}
	}
}

int main() {
	// cout<<"Hello world !";
	int n, m;
	cin >> n >> m;
	vector<vector<int>> edges(n);
	for (int i = 0; i < m; i++) {
		int u, v;
		cin >> u >> v;
		edges[u].push_back(v);
		edges[v].push_back(u);
	}
	int startNode = 0;
	bfs(edges, startNode);
}

// 5 4
// 0 1
// 1 2
// 1 3
// 0 4